src/main/java/org/alliancegenome/curation_api/jobs/executors/HTPExpressionDatasetAnnotationExecutor.java

```diff
@@ -47,43 +49,73 @@ public void execLoad(BulkLoadFileHistory bulkLoadFileHistory) {
+			if (success) {
+				runCleanup(htpExpressionDatasetAnnotationService, bulkLoadFileHistory, dataProvider.name(), previousIds, htpAnnotationsIdsLoaded, fms.getFmsDataType());
+			}
+			bulkLoadFileHistory.finishLoad();
+
 			updateHistory(bulkLoadFileHistory);
+			updateExceptions(bulkLoadFileHistory);
 
-			boolean result = runLoaddatasetid(externalDataBaseEntityService, bulkLoadFileHistory, dataProvider, htpExpressionDatasetData.getData(), datasetIdsLoaded, false);
 		} catch (Exception e) {
 			failLoad(bulkLoadFileHistory, e);
 			e.printStackTrace();
 		}
 	}
 
-	private boolean runLoaddatasetid(ExternalDataBaseEntityService externalDataBaseEntityService, BulkLoadFileHistory bulkLoadFileHistory, BackendBulkDataProvider dataProvider, List<HTPExpressionDatasetAnnotationFmsDTO> htpDatasetData, List<Long> datasetIdsLoaded, boolean isUpdate) {
+	private boolean runLoad(BulkLoadFileHistory history, BackendBulkDataProvider dataProvider, List<HTPExpressionDatasetAnnotationFmsDTO> htpDatasetAnnotations, List<Long> htpAnnotationsIdsLoaded) {
 		ProcessDisplayHelper ph = new ProcessDisplayHelper();
 		ph.addDisplayHandler(loadProcessDisplayService);
-		ph.startProcess("External Database Entity DTO Update for " + dataProvider.name(), htpDatasetData.size());
-		for (HTPExpressionDatasetAnnotationFmsDTO dto : htpDatasetData) {
+		ph.startProcess("HTP Expression Dataset Annotation DTO Update for " + dataProvider.name(), htpDatasetAnnotations.size() * 2);
+
+		updateHistory(history);
+		for (HTPExpressionDatasetAnnotationFmsDTO dto : htpDatasetAnnotations) {
 			try {
-				ExternalDataBaseEntity dbObject = externalDataBaseEntityService.upsert(dto.getDatasetId(), dataProvider);
-				bulkLoadFileHistory.incrementCompleted();
-				if (datasetIdsLoaded != null) {
-					datasetIdsLoaded.add(dbObject.getId());
+				HTPExpressionDatasetAnnotation dbObject = htpExpressionDatasetAnnotationService.upsert(dto, dataProvider);
+				history.incrementCompleted();
+				if (dbObject != null) {
+					htpAnnotationsIdsLoaded.add(dbObject.getId());
 				}
 			} catch (ObjectUpdateException e) {
-				bulkLoadFileHistory.incrementFailed();
-				addException(bulkLoadFileHistory, e.getData());
+				history.incrementFailed();
+				addException(history, e.getData());
 			} catch (Exception e) {
-				bulkLoadFileHistory.incrementFailed();
-				addException(bulkLoadFileHistory, new ObjectUpdateExceptionData(dto, e.getMessage(), e.getStackTrace()));
+				e.printStackTrace();
+				history.incrementFailed();
+				addException(history, new ObjectUpdateExceptionData(dto, e.getMessage(), e.getStackTrace()));
+			}
+			if (history.getErrorRate() > 0.25) {
+				Log.error("Failure Rate > 25% aborting load");
+				updateHistory(history);
+				updateExceptions(history);
+				failLoadAboveErrorRateCutoff(history);
 			}
+			ph.progressProcess();
 		}
-		updateHistory(bulkLoadFileHistory);
-		updateExceptions(bulkLoadFileHistory);
+		updateHistory(history);
```

Comment:
You will need the updateExceptions here also.

Reply:
Sure, will add it